File: frontend/lib/api/radiografias.ts
const API_URL = '/api/proxy'

function authHeaders(token: string) {
  return { Authorization: `Bearer ${token}` }
}

async function apiFetch<T>(path: string, token: string, opts?: RequestInit): Promise<T> {
  const res = await fetch(`${API_URL}${path}`, {
    ...opts,
    headers: { ...authHeaders(token), ...(opts?.headers ?? {}) },
    cache: 'no-store',
  })
  if (!res.ok) {
    const err = await res.json().catch(() => ({}))
    throw new Error((err as { detail?: string }).detail ?? `Error ${res.status}`)
  }
  return res.json()
}

export type TipoRadiografia = 'panoramica' | 'periapical' | 'bitewing' | 'lateral' | 'cefalometrica' | 'otra'

export interface Radiografia {
  id: number
  paciente_id: number
  consultorio_id: number
  archivo_url: string
  tipo: TipoRadiografia
  fecha_toma: string | null
  notas: string | null
  uploaded_by: string | null
  activa: boolean
  created_at: string
  usuarios?: { id: string; nombre: string } | null
}

export async function listarRadiografias(token: string, pacienteId: number): Promise<Radiografia[]> {
  return apiFetch(`/radiografias/admin?paciente_id=${pacienteId}`, token)
}

export async function subirRadiografia(
  token: string,
  data: { paciente_id: number; tipo: TipoRadiografia; fecha_toma?: string; notas?: string; archivo: File },
): Promise<Radiografia> {
  const fd = new FormData()
  fd.append('paciente_id', String(data.paciente_id))
  fd.append('tipo', data.tipo)
  if (data.fecha_toma) fd.append('fecha_toma', data.fecha_toma)
  if (data.notas) fd.append('notas', data.notas)
  fd.append('archivo', data.archivo)

  const res = await fetch(`${API_URL}/radiografias/admin`, {
    method: 'POST',
    headers: authHeaders(token),
    body: fd,
  })
  if (!res.ok) {
    const err = await res.json().catch(() => ({}))
    throw new Error((err as { detail?: string }).detail ?? 'Error al subir')
  }
  return res.json()
}

export async function eliminarRadiografia(token: string, id: number): Promise<void> {
  await apiFetch(`/radiografias/admin/${id}`, token, { method: 'DELETE' })
}

export async function misRadiografias(token: string): Promise<Radiografia[]> {
  return apiFetch('/radiografias/paciente', token)
}

File: backend/app/routers/radiografias.py
"""
M6 — Radiografías e imágenes médicas asociadas al historial clínico.

Tipos: panoramica, periapical, bitewing, lateral, cefalometrica, otra.

Acceso:
- Admin/odontologo: ve/sube/elimina radiografías de pacientes del consultorio
- Recepcionista: NO accede
- Paciente con OTP: ve SUS radiografías (activas)
"""
from __future__ import annotations

import uuid
from typing import Optional

from fastapi import APIRouter, Depends, File, Form, HTTPException, Request, UploadFile
from supabase import Client

from app.db.client import get_supabase_client
from app.routers.auth import require_paciente, require_staff_context
from app.services.audit import log_action

router = APIRouter(prefix="/radiografias", tags=["radiografias"], redirect_slashes=False)

ROLES_RADIOGRAFIA = ("admin", "odontologo", "superadmin")
TIPOS_VALIDOS = {"panoramica", "periapical", "bitewing", "lateral", "cefalometrica", "otra"}
_ALLOWED_EXTS = {"jpg", "jpeg", "png", "webp", "pdf", "dcm"}
_MAX_BYTES = 15 * 1024 * 1024  # 15 MB — radiografías pueden ser grandes


def get_db() -> Client:
    return get_supabase_client()


# ── Admin ────────────────────────────────────────────────────────────────────

@router.get("/admin")
async def listar_radiografias_paciente(
    paciente_id: int,
    ctx: dict = Depends(require_staff_context),
    db: Client = Depends(get_db),
):
    """Lista radiografías activas de un paciente."""
    if ctx["rol"] not in ROLES_RADIOGRAFIA:
        raise HTTPException(status_code=403, detail="Sin permisos para ver radiografías")

    pac = db.table("pacientes").select("id, consultorio_id").eq("id", paciente_id).single().execute()
    if not pac.data:
        raise HTTPException(status_code=404, detail="Paciente no encontrado")
    if not ctx["es_superadmin"] and pac.data["consultorio_id"] != ctx["consultorio_id"]:
        raise HTTPException(status_code=403, detail="Paciente de otro consultorio")

    res = (
        db.table("radiografias")
        .select("*, usuarios:uploaded_by(id, nombre)")
        .eq("paciente_id", paciente_id)
        .eq("activa", True)
        .order("created_at", desc=True)
        .execute()
    )
    return res.data or []


@router.post("/admin")
async def subir_radiografia(
    request: Request,
    paciente_id: int = Form(...),
    tipo: str = Form("panoramica"),
    fecha_toma: Optional[str] = Form(default=None),
    notas: Optional[str] = Form(default=None),
    archivo: UploadFile = File(...),
    ctx: dict = Depends(require_staff_context),
    db: Client = Depends(get_db),
):
    """Sube una radiografía al bucket `radiografias` y crea el registro."""
    if ctx["rol"] not in ROLES_RADIOGRAFIA:
        raise HTTPException(status_code=403, detail="Sin permisos para subir radiografías")

    if tipo not in TIPOS_VALIDOS:
        raise HTTPException(status_code=400, detail=f"Tipo inválido. Opciones: {sorted(TIPOS_VALIDOS)}")

    pac = db.table("pacientes").select("id, consultorio_id").eq("id", paciente_id).single().execute()
    if not pac.data:
        raise HTTPException(status_code=404, detail="Paciente no encontrado")
    if not ctx["es_superadmin"] and pac.data["consultorio_id"] != ctx["consultorio_id"]:
        raise HTTPException(status_code=403, detail="Paciente de otro consultorio")

    contenido = await archivo.read()
    if len(contenido) > _MAX_BYTES:
        raise HTTPException(status_code=413, detail="Archivo demasiado grande (máx 15 MB)")
    ext = (archivo.filename or "file").rsplit(".", 1)[-1].lower() if archivo.filename and "." in archivo.filename else ""
    if ext not in _ALLOWED_EXTS:
        raise HTTPException(
            status_code=400,
            detail=f"Extensión no permitida. Soportadas: {', '.join(sorted(_ALLOWED_EXTS))}",
        )

    consultorio_id = pac.data["consultorio_id"]
    path = f"{consultorio_id}/{paciente_id}/{uuid.uuid4()}.{ext}"

    try:
        db.storage.from_("radiografias").upload(
            path,
            contenido,
            {"content-type": archivo.content_type or "application/octet-stream"},
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error al subir: {e}")

    archivo_url = db.storage.from_("radiografias").get_public_url(path)

    res = db.table("radiografias").insert({
        "paciente_id": paciente_id,
        "consultorio_id": consultorio_id,
        "archivo_url": archivo_url,
        "tipo": tipo,
        "fecha_toma": fecha_toma,
        "notas": notas,
        "uploaded_by": ctx["usuario_id"],
    }).execute()

    log_action(
        consultorio_id=consultorio_id,
        usuario_id=ctx["usuario_id"],
        paciente_id=paciente_id,
        accion="upload_radiografia",
        recurso_tipo="radiografia",
        recurso_id=res.data[0]["id"] if res.data else None,
        request=request,
        metadata={"tipo": tipo, "tamano_bytes": len(contenido)},
    )

    return res.data[0] if res.data else {}


@router.delete("/admin/{radiografia_id}")
async def eliminar_radiografia(
    radiografia_id: int,
    request: Request,
    ctx: dict = Depends(require_staff_context),
    db: Client = Depends(get_db),
):
    """Soft delete: marca `activa=False`. El archivo en Storage queda para auditoría."""
    if ctx["rol"] not in ROLES_RADIOGRAFIA:
        raise HTTPException(status_code=403, detail="Sin permisos")

    r = db.table("radiografias").select("consultorio_id, paciente_id").eq("id", radiografia_id).single().execute()
    if not r.data:
        raise HTTPException(status_code=404, detail="Radiografía no encontrada")
    if not ctx["es_superadmin"] and r.data["consultorio_id"] != ctx["consultorio_id"]:
        raise HTTPException(status_code=403, detail="Radiografía de otro consultorio")

    db.table("radiografias").update({"activa": False}).eq("id", radiografia_id).execute()

    log_action(
        consultorio_id=r.data["consultorio_id"],
        usuario_id=ctx["usuario_id"],
        paciente_id=r.data["paciente_id"],
        accion="delete_radiografia",
        recurso_tipo="radiografia",
        recurso_id=radiografia_id,
        request=request,
    )
    return {"ok": True}


# ── Paciente (JWT OTP) ───────────────────────────────────────────────────────

@router.get("/paciente")
async def mis_radiografias(
    payload: dict = Depends(require_paciente),
    db: Client = Depends(get_db),
):
    """Paciente ve sus propias radiografías activas."""
    paciente_id = int(payload["sub"])
    res = (
        db.table("radiografias")
        .select("id, archivo_url, tipo, fecha_toma, notas, created_at")
        .eq("paciente_id", paciente_id)
        .eq("activa", True)
        .order("created_at", desc=True)
        .execute()
    )
    return res.data or []
